--- cyrus-imapd-2.4.4/sieve/Makefile.in~	2010-11-12 00:15:33.000000000 +0200
+++ cyrus-imapd-2.4.4/sieve/Makefile.in	2010-11-26 18:36:31.623192952 +0200
@@ -127,7 +127,7 @@
 	$(CC) $(LDFLAGS) -o sievec sievec.o libsieve.a $(DEPLIBS) $(LIBS)
 
 sieved: sieved.o libsieve.a
-	$(CC) $(LDFLAGS) -o sieved sieved.o libsieve.a $(DEPLIBS) $(LIBS)
+	$(CC) $(LDFLAGS) -o sieved sieved.o libsieve.a $(LIBS) $(DEPLIBS)
 
 
 sieve_interface.h: sieve_err.h
