--- cyrus-imapd-2.0.16/imap/imapparse.c.orig	Mon Jul 16 21:31:14 2001
+++ cyrus-imapd-2.0.16/imap/imapparse.c	Thu Dec  5 11:53:09 2002
@@ -52,7 +52,7 @@
 	       struct buf *buf, int type)
 {
     int c;
-    int i, len = 0;
+    unsigned int i, len = 0;
     int sawdigit = 0;
     int isnowait;
 
@@ -168,6 +168,16 @@
 	    if (c != EOF) prot_ungetc(c, pin);
 	    return EOF;
 	}
+	if (len > 65536) {
+	    if (isnowait) {
+		for (i = 0; i < len; i++)
+		    c = prot_getc(pin);
+	    }
+	    prot_printf(pout, "* BAD Literal too large\r\n");
+	    prot_flush(pout);
+	    if (c != EOF) prot_ungetc(c, pin);
+	    return EOF;
+	}
 	if (len >= buf->alloc) {
 	    buf->alloc = len+1;
 	    buf->s = xrealloc(buf->s, buf->alloc+1);
