diff -urNp -x '*.orig' cyrus-imapd-3.0.9.org/lib/util.c cyrus-imapd-3.0.9/lib/util.c
--- cyrus-imapd-3.0.9.org/lib/util.c	2019-03-15 01:31:20.000000000 +0100
+++ cyrus-imapd-3.0.9/lib/util.c	2021-05-07 22:56:04.721472003 +0200
@@ -579,7 +579,7 @@ EXPORTED int set_caps(int stage __attrib
 }
 #endif
 
-static int cap_setuid(int uid, int is_master)
+static int cyrus_cap_setuid(int uid, int is_master)
 {
     int r;
 
@@ -599,7 +599,7 @@ EXPORTED int become_cyrus(int is_master)
     int result;
     static uid_t uid = 0;
 
-    if (uid) return cap_setuid(uid, is_master);
+    if (uid) return cyrus_cap_setuid(uid, is_master);
 
     const char *cyrus = cyrus_user();
     const char *mail = cyrus_group();
@@ -645,7 +645,7 @@ EXPORTED int become_cyrus(int is_master)
         return -1;
     }
 
-    result = cap_setuid(newuid, is_master);
+    result = cyrus_cap_setuid(newuid, is_master);
 
     /* Only set static uid if successful, else future calls won't reset gid */
     if (result == 0)
diff -urNp -x '*.orig' cyrus-imapd-3.0.9.org/master/master.c cyrus-imapd-3.0.9/master/master.c
--- cyrus-imapd-3.0.9.org/master/master.c	2019-03-15 01:31:20.000000000 +0100
+++ cyrus-imapd-3.0.9/master/master.c	2021-05-07 22:56:04.721472003 +0200
@@ -285,7 +285,7 @@ static void get_statsock(int filedes[2])
         fatalf(1, "unable to set close-on-exec: %m");
 }
 
-static int cap_bind(int socket, struct sockaddr *addr, socklen_t length)
+static int cyrus_cap_bind(int socket, struct sockaddr *addr, socklen_t length)
 {
     int r;
 
@@ -643,7 +643,7 @@ static void service_create(struct servic
 #endif
 
         oldumask = umask((mode_t) 0); /* for linux */
-        r = cap_bind(s->socket, res->ai_addr, res->ai_addrlen);
+        r = cyrus_cap_bind(s->socket, res->ai_addr, res->ai_addrlen);
         umask(oldumask);
         if (r < 0) {
             syslog(LOG_ERR, "unable to bind to %s/%s socket: %m",
