--- cyrus-imapd-3.0.9/imap/cyr_virusscan.c.orig	2019-03-15 01:31:20.000000000 +0100
+++ cyrus-imapd-3.0.9/imap/cyr_virusscan.c	2019-04-16 18:46:50.316701915 +0200
@@ -126,6 +126,11 @@
 /* ClamAV implementation */
 #include <clamav.h>
 
+#ifndef __CLAMAV_CONFIG_H
+/* clamav 0.101.2 dropped installing clamav-config.h, assume it's major 9 for now */
+#define LIBCLAMAV_MAJORVER 9
+#endif
+
 struct clamav_state {
     struct cl_engine *av_engine;
 };
