--- cyrus-imapd-2.0.14/configure.in.wiget	Fri Jul  6 18:55:49 2001
+++ cyrus-imapd-2.0.14/configure.in	Fri Jul  6 18:56:39 2001
@@ -344,10 +344,10 @@
 fi
 
 AC_SUBST(WITH_MAP)
-AC_ARG_WITH(lock,
+AC_ARG_WITH([lock],
   [  --with-lock=METHOD      force use of METHOD for locking (flock or fcntl)],
-  WITH_LOCK="$withval",
-  AC_CHECK_FUNC(flock,WITH_LOCK="flock",WITH_LOCK="fcntl"))
+  [WITH_LOCK="$withval"],
+  [AC_CHECK_FUNC([flock],[WITH_LOCK="flock"],[WITH_LOCK="fcntl"])])
 AC_SUBST(WITH_LOCK)
 
 dnl for makedepend and AFS.
