--- cyrus-imapd-2.0.16/imap/fud.c.orig	Tue Dec  3 10:38:01 2002
+++ cyrus-imapd-2.0.16/imap/fud.c	Thu Dec  5 12:04:22 2002
@@ -313,7 +313,7 @@
             sendto(soc,"PERMDENY",9,0,(struct sockaddr *) &sfrom, sizeof(sfrom));       
             break;
         case REQ_OK:
-            siz = sprintf(buf,"%s|%s|%d|%d|%d",user,mbox,numrecent,(int) lastread,(int) lastarrived);
+            siz = snprintf(buf,sizeof(buf),"%s|%s|%d|%d|%d",user,mbox,numrecent,(int) lastread,(int) lastarrived);
             sendto(soc,buf,siz,0,(struct sockaddr *) &sfrom, sizeof(sfrom));       
             break;
         case REQ_UNK:
