--- cyrus-imapd-2.0.16/configure.in.orig	Mon Jul 29 12:36:24 2002
+++ cyrus-imapd-2.0.16/configure.in	Mon Jul 29 12:55:57 2002
@@ -150,8 +150,8 @@
 	AC_CHECK_LIB(fl,main)
 
  	AC_SEARCH_LIBS(regcomp, rx regex,
- 			AC_DEFINE(ENABLE_REGEX)
-		 	AC_CHECK_HEADER(rxposix.h, AC_DEFINE(HAVE_RX)))
+ 			[AC_DEFINE(ENABLE_REGEX)
+		 	AC_CHECK_HEADER(rxposix.h, AC_DEFINE(HAVE_RX))])
   		
 
 	PRE_SUBDIRS="${PRE_SUBDIRS} $sievedir"
--- cyrus-imapd-2.0.16/cmulocal/libwrap.m4.orig	Fri Jun  9 21:02:44 2000
+++ cyrus-imapd-2.0.16/cmulocal/libwrap.m4	Mon Jul 29 13:07:21 2002
@@ -12,7 +12,7 @@
     fi
     cmu_save_LIBS="$LIBS"
     AC_CHECK_LIB(wrap, request_init,
-		 AC_CHECK_HEADER(tcpd.h,, with_libwrap=no),
+		 [AC_CHECK_HEADER(tcpd.h,, with_libwrap=no)],
 		 with_libwrap=no, ${LIB_SOCKET})
     LIBS="$cmu_save_LIBS"
   fi
--- cyrus-imapd-2.0.16/cmulocal/ucdsnmp.m4.orig	Sun Nov  5 23:11:28 2000
+++ cyrus-imapd-2.0.16/cmulocal/ucdsnmp.m4	Mon Jul 29 13:09:08 2002
@@ -12,7 +12,7 @@
     fi
     cmu_save_LIBS="$LIBS"
     AC_CHECK_LIB(snmp, sprint_objid,
-		 AC_CHECK_HEADER(ucd-snmp/version.h,, with_ucdsnmp=no),
+		 [AC_CHECK_HEADER(ucd-snmp/version.h,, with_ucdsnmp=no)],
 		 with_ucdsnmp=no, ${LIB_SOCKET})
     LIBS="$cmu_save_LIBS"
   fi
