diff -burN cyrus-imapd-2.1.14-orig/configure.in cyrus-imapd-2.1.14/configure.in
--- cyrus-imapd-2.1.14-orig/configure.in	2003-09-04 13:29:51.877651520 +0200
+++ cyrus-imapd-2.1.14/configure.in	2003-09-04 13:32:18.193408144 +0200
@@ -793,7 +793,8 @@
       ;;
   *) # use whatever they told us, or whatever we found
      COMPILE_ET="${with_com_err}/bin/compile_et"
-     COM_ERR_LIBS="${with_com_err}/lib/libcom_err.a"
+     COM_ERR_LDFLAGS="-L${with_com_err}/lib/libcom_err.a"
+     COM_ERR_LIBS="-lcom_err"
      COM_ERR_CPPFLAGS="-I${with_com_err}/include"
      # Ever get the feeling people hide this stuff on purpose?
      if test -d "${with_com_err}/include/et" ; then
