--- cyrus-imapd-2.3.8/lib/cyrusdb_berkeley.c.wig	2007-02-18 22:57:48.000000000 +0100
+++ cyrus-imapd-2.3.8/lib/cyrusdb_berkeley.c	2007-02-18 23:07:06.000000000 +0100
@@ -176,10 +176,10 @@
 	syslog(LOG_WARNING,
 	       "DBERROR: invalid berkeley_locks_max value, using internal default");
     } else {
-	r = dbenv->set_lk_max(dbenv, opt);
+	r = dbenv->set_lk_max_locks(dbenv, opt);
 	if (r) {
-	    dbenv->err(dbenv, r, "set_lk_max");
-	    syslog(LOG_ERR, "DBERROR: set_lk_max(): %s", db_strerror(r));
+	    dbenv->err(dbenv, r, "set_lk_max_locks");
+	    syslog(LOG_ERR, "DBERROR: set_lk_max_locks(): %s", db_strerror(r));
 	    abort();
 	}
     }
