--- cyrus-imapd-2.0.5/man/Makefile.in	Sat Jun 17 18:49:10 2000
+++ cyrus-imapd-2.0.5/man/Makefile.in	Sat Jun 17 18:49:48 2000
@@ -60,6 +60,7 @@
 prefix = @prefix@
 exec_prefix = @exec_prefix@
 cyrus_prefix = @cyrus_prefix@
+mandir=@mandir@
 
 MAN1 = $(srcdir)/cyradm.1 $(srcdir)/imtest.1 $(srcdir)/installsieve.1
 MAN3 = $(srcdir)/imclient.3
@@ -75,19 +76,19 @@
 install:
 	for file in $(MAN1); \
 	do \
-		$(INSTALL) -m 644 $$file $(DESTDIR)$(prefix)/man/man1 || exit 1; \
+		$(INSTALL) -m 644 $$file $(DESTDIR)$(mandir)/man1 || exit 1; \
 	done
 	for file in $(MAN3); \
 	do \
-		$(INSTALL) -m 644 $$file $(DESTDIR)$(prefix)/man/man3 || exit 1; \
+		$(INSTALL) -m 644 $$file $(DESTDIR)$(mandir)/man3 || exit 1; \
 	done
 	for file in $(MAN5); \
 	do \
-		$(INSTALL) -m 644 $$file $(DESTDIR)$(prefix)/man/man5 || exit 1; \
+		$(INSTALL) -m 644 $$file $(DESTDIR)$(mandir)/man5 || exit 1; \
 	done
 	for file in $(MAN8); \
 	do \
-		$(INSTALL) -m 644 $$file $(DESTDIR)$(prefix)/man/man8 || exit 1; \
+		$(INSTALL) -m 644 $$file $(DESTDIR)$(mandir)/man8 || exit 1; \
 	done
 
 .c.o:
--- cyrus-imapd-2.0.5/Makefile.in	Sat Jun 17 18:50:20 2000
+++ cyrus-imapd-2.0.5/Makefile.in	Sat Jun 17 18:50:46 2000
@@ -56,6 +56,7 @@
 prefix = @prefix@
 exec_prefix = @exec_prefix@
 cyrus_prefix = @cyrus_prefix@
+mandir = @mandir@
 
 all:: checkdepend
 	@for d in  $(SUBDIRS); \
@@ -68,10 +69,10 @@
 	$(srcdir)/install-sh -d ${DESTDIR}$(exec_prefix)/bin
 	$(srcdir)/install-sh -d ${DESTDIR}$(exec_prefix)/lib
 	$(srcdir)/install-sh -d ${DESTDIR}$(prefix)/include/cyrus
-	$(srcdir)/install-sh -d ${DESTDIR}$(prefix)/man/man1
-	$(srcdir)/install-sh -d ${DESTDIR}$(prefix)/man/man3
-	$(srcdir)/install-sh -d ${DESTDIR}$(prefix)/man/man5
-	$(srcdir)/install-sh -d ${DESTDIR}$(prefix)/man/man8
+	$(srcdir)/install-sh -d ${DESTDIR}$(mandir)/man1
+	$(srcdir)/install-sh -d ${DESTDIR}$(mandir)/man3
+	$(srcdir)/install-sh -d ${DESTDIR}$(mandir)/man5
+	$(srcdir)/install-sh -d ${DESTDIR}$(mandir)/man8
 	@for d in  $(SUBDIRS); \
 	do \
 		(cd $$d; echo "### Making" install "in" `pwd`;	\
