--- cyrus-imapd-3.8.1/imap/vcard_support.c~	2023-09-11 03:23:52.000000000 +0200
+++ cyrus-imapd-3.8.1/imap/vcard_support.c	2023-11-26 23:16:21.657955478 +0100
@@ -42,6 +42,7 @@
  */
 
 #include <config.h>
+#include <libxml/parser.h>
 #include <libxml/tree.h>
 
 #include "vcard_support.h"
--- cyrus-imapd-3.8.1/imap/http_dav.c~	2023-09-11 03:23:52.000000000 +0200
+++ cyrus-imapd-3.8.1/imap/http_dav.c	2023-11-26 23:18:32.021051917 +0100
@@ -93,6 +93,7 @@
 #include "imap/imap_err.h"
 
 #include <errno.h>
+#include <libxml/parser.h>
 #include <libxml/uri.h>
 
 static const struct dav_namespace_t {
--- cyrus-imapd-3.8.1/imap/http_dav_sharing.c~	2023-09-11 03:23:52.000000000 +0200
+++ cyrus-imapd-3.8.1/imap/http_dav_sharing.c	2023-11-26 23:19:26.519004807 +0100
@@ -59,6 +59,8 @@
 #include "imap/http_err.h"
 #include "imap/imap_err.h"
 
+#include <libxml/parser.h>
+
 #define DAVNOTIFICATION_CONTENT_TYPE \
     "application/davnotification+xml; charset=utf-8"
 
--- cyrus-imapd-3.8.1/imap/xcal.c~	2023-09-11 02:04:18.000000000 +0200
+++ cyrus-imapd-3.8.1/imap/xcal.c	2023-11-26 23:19:52.066272967 +0100
@@ -47,6 +47,7 @@
 #include <stddef.h> /* for offsetof() macro */
 #include <syslog.h>
 
+#include <libxml/parser.h>
 #include <libxml/tree.h>
 
 #include "httpd.h"
