--- cyrus-imapd-3.0.9/imap/httpd.c~	2019-03-15 01:31:20.000000000 +0100
+++ cyrus-imapd-3.0.9/imap/httpd.c	2021-10-02 21:53:14.171936180 +0200
@@ -914,7 +914,7 @@
                cyrus_version(),
                SASL_VERSION_MAJOR, SASL_VERSION_MINOR, SASL_VERSION_STEP);
 #ifdef HAVE_SSL
-    buf_printf(&serverinfo, " OpenSSL/%s", SHLIB_VERSION_NUMBER);
+    buf_printf(&serverinfo, " OpenSSL/%s", OPENSSL_SHLIB_VERSION);
 #endif
 
 #ifdef HAVE_NGHTTP2
